### PR TITLE
Refactor Symbol::ListInputs to be static

### PR DESCRIPTION
**Pull request by larroy**
This avoid needing to create an unused Symbol when needing to filter Nodes by Input Option.

For example here: https://github.com/apache/incubator-mxnet/blob/master/src/imperative/imperative.cc#L321

This code in MXNet is being refactored and depends on this refactor.

https://github.com/larroy/mxnet/tree/fc_higher_order_grad_2

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

**Comment by tqchen** [comment]
Symbol itself is simply the same as vector<NodePtr> so likely it is going to be the same

**Comment by larroy** [comment]
a function that doesn't depend on the object is better static, this avoid needs to create a Symbol in the backward pass in MXNet just to call this function. For NNVM is no change as there's a member calling the static, so API remains the same. Any concerns for merging @tqchen ?

**Comment by tqchen** [comment]
As I said in the last thread, creating a new Symbol won't be much of a problem, because structurally it is equivalent to ```vector<NodeEntry>```. It will also make the python interfacing easy, as Symbol itself is exposed in python. So there won't be too much performance gains you will get by making this change.

**Comment by larroy** [comment]
Is not about performance, it's more about semantics, there's no need to create a symbol when one wants to list the inputs from a graph. Did you check the link to the MXNet code that I was referring to? It's about interface segregation as in https://en.wikipedia.org/wiki/SOLID

**Comment by tqchen** [comment]
I believe this is a nit change that could go either way. 

The main problem I have about introducing the new API is will results in other questions, e.g. do we want to also add static APIs for other functions like ListInputNames? Should these static functions belong to Symbol at all given they are only related to nodes? 

As a lazy solution, I would rather avoid these questions by just creating a Symbol and call ListInputs, which is just fine.

**Comment by tqchen** [comment]
This being said, if we just stop at adding this function, and add sufficient document to it, I am fine merging it in.

**Comment by larroy** [comment]
For the backward pass refactor that I'm working on I would only need this one change. Could you suggest a more appropiate channel for these discussions should these future questions arise?  Slack, MXnet dev or tvm dev?

**Comment by tqchen** [comment]
I think the discussion can happens fine as long as it is in the public channels.  This is an API that could be used either way and won’t block your code change. 

I have already listed my technical reasonings. As I said, I am open to merge it in but there are also reason why not to, so please at least discuss these points

**Comment by larroy** [comment]
Thanks for your comments and open mind.

My approach to this PR is from a user of NNVM API coming from its usage in MXNet. I don't want to create an unnecessary object. I looked at this method and it should be static as it doesn't use any member variables to function, this for me is a basic rule of thumb of software engineering, that's pretty much all there is to this PR intention.\

It avoids the creation of this additional object that is not needed. The gradient function is operating on Node and NodePtr, NodeEntry etc as you are well aware. So from a user perspective yes I would like this merged or factored out of this class if you think there's a better place to put this kind of functions, please make a suggestion.

If your request is that I also make ListInputNames or any other similar function static for consistency I'm fine with adding some additional work to keep everything consistent, as I think this would be a fair request.

Regarding discussion, what I was alluding to is a more general and extended discussion about improving the nnvm graph semantics, for example the graph and nodes could be more generic and have attributes as template arguments (think of NodeAttributes). This would be better than we currently have, for example autograd is shoved into the "info" field which is of type "any", you end up losing type information. This is beyond the scope of this PR.

Please let me know how we should move forward, I obviously would like this PR to be merged otherwise I wouldn't have opened it. I'm open to have additional refactorings requested. If it's not merged I will create a symbol as before but I would prefer to avoid that, for me is just a small ramification of a larger refactor.

Thanks.

**Comment by tqchen** [comment]
OK let me try again to summarize the discussions here. I think the API style really has things to do with a different perspective of what defines the graph. 

- The original API's view says that the Symbol object itself defines the graph, and whenever you wanted to create ListInputs, create a temporary Symbol object to do so.
- Your current view says that Graph itself should not be bound to Symbol, and vector<NodeEntry> should be sufficient to define a graph, and thus the static function.

Both ways are valid engineering points. I think I could have go with either one. But only changing one API ListInputNodes, represent a breach in terms of the inconsistency of the design -- because of parts of the API goes into one style while rest remains in another.

I personally know very where when should we use static method vs member functions, even given my pedantic view about the code and infra, I do believe this is a case that could have gone either way. 

Given that the historical design has already made an opinionated view about what a symbolic graph should be, I would prefer to keep the API as it is. Creating a temporary object is fine, as one-liner and wouldn't breach anything in the code and it keeps the view of API consistent with the original design. It won't be a critical blocker for your changes rather than the different design choice.

While I understand sometimes different developer has different points of views and I fully understand potential frustration you might have given your take of design(which was apparently quite clear). Please think a bit more again about the technical points here and see if it the other side of the argument also makes sense.

If I would have to make a recommendation. I would recommend just create a temporary symbol and keep the same perspective as the original design.

**Comment by larroy** [comment]
I disagree. That means doing an additional copy of the nodes for no reason. Also we are using Node objects constantly, not Symbol.

This is how the refactored code looks like:
 https://github.com/larroy/mxnet/blob/fc_higher_order_grad_2/src/imperative/imperative.cc#L345

About Graph being bound to Symbol, I don't have a strong opinion since I didn't write the original code myself. Actually the change that introduced this call was reviewed by you, so I assume you were supportive of using the graph through Node and NodePtrs and not Symbol:  

https://github.com/apache/incubator-mxnet/pull/7698

**Comment by tqchen** [comment]
About Graph being bound to Symbol. I am specifically referring to the ListXYZ API in the Symbol, as all of them now are member functions and treat general List Input manipulation as a special operation to the data structure defined by the symbol.

Performance-wise, the additional copy won't become the bottleneck of your code given that there are already many vector and additional data structure created during PostOrder Visit. And creating the symbol will just be fine. The reason for doing so is to respect the original semantics of ListXYZ as they are bound to the Symbol. 

I think at the moment all the argument has been presented. I have read your refactored code(fully understand what you want and your arguments),  and have presented my recommendation :) This is certainly not perfect,  as everything is part of the tradeoff in software. Just like there are also additional trade-offs you make by using template vs type-erased any (which we could discuss once you brought up the RFC).

I do hope that you can also take a look at the discussion thread and consider it for a while. To me, the code itself is not a big deal, but I do hope we can handle technical debate in a diplomatic fashion.

**Comment by larroy** [comment]
@tqchen thanks for your responses. If possible I would appreciate to have this merged. I think for the moment I don't need more changes in TVM / NNVM for my refactor in MXNet so it would help. I believe the code is improved by this PR, and doesn't make a difference for the existing API. I don't have anything else to add to the discussion. I guess it only remains for you to take a final decision to merge or close the PR. Thanks for constructive the discussion.

**Comment by tqchen** [comment]
I will close this PR for now due to the reason listed. Thanks @larroy for raising the issues and constructive discussions. Hopefully, it won't block you technically to implement the features you need
